cli/src/cmdtree.rs

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Open Network Fabric Authors
+
+//! Defines a command tree of Nodes
+
+use colored::Colorize;
+use std::collections::BTreeMap;
+use std::collections::VecDeque;
+
+#[derive(Default)]
+pub struct NodeArg {
+    pub name: String,
+    pub choices: Vec<String>,
+}
+#[allow(unused)]
+impl NodeArg {
+    pub fn new(name: &str) -> Self {
+        Self {
+            name: name.to_owned(),
+            choices: Vec::new(),
+        }
+    }
+    pub fn choice(mut self, choice: &str) -> Self {
+        self.choices.push(choice.to_owned());
+        self
+    }
+}
+
+#[derive(Default)]
+pub struct Node {
+    pub(crate) name: String,
+    pub depth: u16,
+    pub children: BTreeMap<String, Node>,
+    pub(crate) description: Option<&'static str>,
+    pub action: Option<u16>,
+    pub(crate) args: Vec<NodeArg>,
+    pub(crate) hidden: bool,
+}
+#[allow(unused)]
+impl Node {
+    pub fn new(name: &str) -> Self {
+        Self {
+            name: name.to_owned(),
+            hidden: false,
+            ..Default::default()
+        }
+    }
+    pub fn action(mut self, action: u16) -> Self {
+        self.action = Some(action);
+        self
+    }
+    pub fn arg(mut self, arg: &str) -> Self {
+        self.args.push(NodeArg::new(arg));
+        self
+    }
+    pub fn arg_add(mut self, arg: NodeArg) -> Self {
+        self.args.push(arg);
+        self
+    }
+    pub fn find_arg(&self, name: &str) -> Option<&NodeArg> {
+        self.args.iter().find(|&arg| arg.name == name)
+    }
+    pub fn hidden(mut self) -> Self {
+        self.hidden = true;
+        self
+    }
+    pub fn desc(mut self, description: &'static str) -> Self {
+        self.description = Some(description);
+        self
+    }
+    fn set_depth(&mut self, depth: u16) {
+        self.depth = depth;
+        self.children
+            .values_mut()
+            .for_each(|c| c.set_depth(depth + 1));
+    }
+    fn add(&mut self, mut cmd: Node) {
+        if cmd.name.is_empty() {
+            //if a node has no name, skip it and adopt
+            //its children. This is just to ease the
+            //creation of the tree by creating
+            //unnamed commands that will not show up
```

Comment:
```suggestion
            // If a node has no name, skip it and adopt
            // its children. This is just to ease the
            // creation of the tree by creating
            // unnamed commands that will not show up.
```

Reply:
Doesn't have to be doc comments, we do usually leave a space after the `//` marker - at least I can't find any other comment instance without that space.